### PR TITLE
Cantus large real data sets experiment and doc update

### PR DESCRIPTION
**Pull request by Yueqiao12Zhang**
Uses a real data sets from CSV export directly and perform necessary procedures to convert to RDF.

### CODE REVIEW

File: cantusdb/README.md
Comment (candlecao) [inline cantusdb/README.md:1]: "all the source CSV into the raw/ folder" --where is the raw/ folder?

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@ psqldump
 .DS_Store
 Csv2rdf/out_rdf.ttl
 .history/*
-cantusdb/data/*
+cantusdb/data/raw/*.csv
+Cantusdb/data/deprecated/*
```
Comment (dchiller) [inline .gitignore:8]: It looks like a file in this directory made it into the PR? `cantusdb_01102024.csv`

File: cantusdb/data/raw/README.md
Comment (dchiller) [inline cantusdb/data/raw/README.md:1]: I don't think you need to do this...

File: cantusdb/csv/fetch.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+"""
+
+import json
+import os
+import time
+import requests
+
+# URL of the file to download
+SOURCE_PATH = os.path.join(
+    os.path.dirname(__file__), "../data/mappings/sources_short.json"
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:12]: you don't explain anywhere what "sources_short.json" is....

File: cantusdb/csv/fetch.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+"""
+fetch the CSV exports
+"""
+
+import json
+import os
+import time
+import requests
+
+# URL of the file to download
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:10]: I don't think this comment is correct.

File: cantusdb/csv/fetch.py (hunk 2 above)
Comment (dchiller) [inline cantusdb/csv/fetch.py:12]: you can just use the relative path...

File: cantusdb/csv/fetch.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+SOURCE_PATH = os.path.join(
+    os.path.dirname(__file__), "../data/mappings/sources_short.json"
+)
+with open(SOURCE_PATH, mode="r", encoding="utf-8") as f:
+    SOURCE_LIST = json.load(f)
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:15]: This isn't a constant so I think probably don't use the uppercase convention.

File: cantusdb/csv/fetch.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+# Send a GET request to the URL
+for src in SOURCE_LIST:
+    SRC_URL = "https://cantusdatabase.org/source/" + str(src) + "/csv"
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:19]: Same here with constant naming convention.

File: cantusdb/csv/fetch.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+# URL of the file to download
+SOURCE_PATH = os.path.join(
+    os.path.dirname(__file__), "../data/mappings/sources_short.json"
+)
+with open(SOURCE_PATH, mode="r", encoding="utf-8") as f:
+    SOURCE_LIST = json.load(f)
+
+# Send a GET request to the URL
+for src in SOURCE_LIST:
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:18]:
```suggestion
for source_id in SOURCE_LIST:
```

File: cantusdb/csv/fetch.py (hunk 5 above)
Comment (dchiller) [reply cantusdb/csv/fetch.py:19]: Also, f-string?

File: cantusdb/csv/fetch.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    SRC_URL = "https://cantusdatabase.org/source/" + str(src) + "/csv"
+    response = requests.get(SRC_URL,  timeout=500)
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:20]: 8 minutes seems like a looong timeout...

File: cantusdb/csv/fetch.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
+"""
+fetch the CSV exports
```
Comment (dchiller) [inline cantusdb/csv/fetch.py:2]: What csv exports? From where?

File: cantusdb/csv/merge.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import pandas as pd
+
+INPUT_PATH = os.path.join(os.path.dirname(__file__), "../data/raw")
+OUTPUT_PATH = os.path.join(os.path.dirname(__file__), "../data/cantus.csv")
+MAPPING_PATH = os.path.join(os.path.dirname(__file__), "../data/mappings")
+
+# Create an empty list to store individual DataFrames
+dfs = []
+genre_mappings = pd.read_csv(
+    os.path.join(MAPPING_PATH, "genres.tsv"), sep="\t", header=0, index_col=False
```
Comment (dchiller) [inline cantusdb/csv/merge.py:15]: Where does `genres.tsv` come from? I don't think it's documented...

File: cantusdb/csv/merge.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+)
+service_mappings = pd.read_csv(
+    os.path.join(MAPPING_PATH, "services.tsv"), sep="\t", header=0, index_col=False
```
Comment (dchiller) [inline cantusdb/csv/merge.py:18]: Where does `services.tsv` come from? I don't think it's documented...

File: cantusdb/csv/merge.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+service_mappings = dict(zip(service_mappings["Service"], service_mappings["Description"]))
+
+# Iterate over all files in the directory
+for filename in os.listdir(INPUT_PATH):
+    # Construct the full file path
+    file_path = os.path.join(INPUT_PATH, filename)
+
+    # Check if the file is a CSV file
+    if filename.endswith(".csv"):
+        # Read the CSV file into a DataFrame
```
Comment (dchiller) [inline cantusdb/csv/merge.py:30]: Could this be a place to use glob module?

File: cantusdb/csv/merge.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        df = pd.read_csv(file_path)
+        df["source_id"] = "https://cantusdatabase.org/sources/" + str(filename.split(".")[0])
```
Comment (dchiller) [inline cantusdb/csv/merge.py:32]: f-string?

File: cantusdb/data/raw/README.md
Comment (Yueqiao12Zhang) [reply cantusdb/data/raw/README.md:1]: If I don't do this, I think if you pull the PR, the raw folder will not appear. Then the program will run into a bug when it's trying to download and write file to the raw folder.

File: .gitignore (hunk 1 above)
Comment (Yueqiao12Zhang) [reply .gitignore:8]: I will remove.

File: cantusdb/csv/fetch.py (hunk 2 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/fetch.py:12]: I always wanted to do this, but I searched and I don't see any better ways of using relative path..

File: cantusdb/csv/fetch.py (hunk 4 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/fetch.py:15]: Pylint tells me to use uppercase

File: cantusdb/csv/merge.py (hunk 11 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/merge.py:30]: Why are we preferring glob instead of os.listdir?

File: cantusdb/csv/fetch.py (hunk 2 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/fetch.py:12]: It's in README.

File: cantusdb/csv/fetch.py (hunk 4 above)
Comment (ahankinson) [reply cantusdb/csv/fetch.py:15]: Pulling is wrong then. What’s the message?

File: cantusdb/data/raw/README.md
Comment (dchiller) [inline cantusdb/data/raw/README.md:1]: I'm still not sure why this is here. If the folder is empty in remote, why does it need to exist in remote?

File: cantusdb/README.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+https://cantusdatabase.org/marginalia
+https://cantusdatabase.org/sequence
+https://cantusdatabase.org/office
+https://cantusindex.org/id
+https://cantusdatabase.org/finalis
+https://cantusdatabase.org/extra
+
+##  3. Abbreviation mappings
+
+Some cells in the original CSV export are in abbreviations. Two TSV files in the data/mappings folder contains the official mappings for the abbreviations in Cantus DB. They are used to map the abbreviations in genres column and services colum to their real literals. This process ease reconciliation.
```
Comment (dchiller) [inline cantusdb/README.md:42]: You could add URLs for where you got these from here...

File: cantusdb/csv/merge.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    os.path.join(MAPPING_PATH, "services.tsv"), sep="\t", header=0, index_col=False
+)
+genre_mappings = dict(zip(genre_mappings["Genre"], genre_mappings["Description"]))
+service_mappings = dict(zip(service_mappings["Service"], service_mappings["Description"]))
+
+# Iterate over all files in the directory
+for filename in glob.glob(f"{INPUT_PATH}/*.csv", recursive=False):
+    # Read the CSV file into a DataFrame
+    df = pd.read_csv(filename)
+    entity_id = (filename.split("/")[-1])[0:6] # get the 6-digit source id
```
Comment (dchiller) [inline cantusdb/csv/merge.py:28]: Sources can have IDs longer than 6 characters (see, for example https://cantusdatabase.org/source/1000019)

File: cantusdb/data/raw/README.md
Comment (Yueqiao12Zhang) [reply cantusdb/data/raw/README.md:1]: All output CSVs will be downloaded here. If other collaborators pull this branch and run the script directly, they don't have this folder, and the script will run into an error.

File: cantusdb/csv/merge.py (hunk 14 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/merge.py:28]: I will try another approach.

File: cantusdb/csv/fetch.py (hunk 4 above)
Comment (Yueqiao12Zhang) [reply cantusdb/csv/fetch.py]: My bad, it does not return the message now, I don't know why.

File: cantusdb/data/raw/README.md
Comment (dchiller) [reply cantusdb/data/raw/README.md:1]: Then either (a) it should be documented that the user needs to create this folder, or (b) the script should check whether this folder exists and if it doesn't, create it.

**Review by dchiller — APPROVED** [review]

**Pull request merged**